File: record_android/android/src/main/java/com/llfbandit/record/record/format/AudioFormats.java
package com.llfbandit.record.record.format;

import android.media.MediaCodecInfo;
import android.media.MediaCodecList;
import android.media.MediaFormat;
import android.os.Build;
import android.util.Log;

public class AudioFormats {
  private AudioFormats() {
  }

  public static boolean isEncoderSupported(String mimeType) {
    MediaCodecList mcl = new MediaCodecList(MediaCodecList.REGULAR_CODECS);

    for (MediaCodecInfo info : mcl.getCodecInfos()) {
      if (info.isEncoder()) {
        for (String supportedType : info.getSupportedTypes()) {
          if (supportedType.equalsIgnoreCase(mimeType)) {
            return true;
          }
        }
      }
    }

    return false;
  }

  public static String getMimeType(String encoder) {
    switch (encoder) {
      case "aacLc":
      case "aacEld":
      case "aacHe":
        return MediaFormat.MIMETYPE_AUDIO_AAC;
      case "amrNb":
        return MediaFormat.MIMETYPE_AUDIO_AMR_NB;
      case "amrWb":
        return MediaFormat.MIMETYPE_AUDIO_AMR_WB;
      case "wav":
      case "pcm16bit":
      case "pcm8bit":
        return MediaFormat.MIMETYPE_AUDIO_RAW;
      case "opus":
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
          return MediaFormat.MIMETYPE_AUDIO_OPUS;
        }

        Log.d("Record", "Api < 29: Opus is not supported. Reverting to default AAC media format.");
        return MediaFormat.MIMETYPE_AUDIO_AAC;
      case "vorbisOgg":
        return MediaFormat.MIMETYPE_AUDIO_VORBIS;
      default:
        return null;
    }
  }
}

File: record_android/android/src/main/kotlin/com/llfbandit/record/record/format/AudioFormats.kt
package com.llfbandit.record.record.format

import android.media.MediaCodecList
import android.media.MediaFormat
import com.llfbandit.record.record.AudioEncoder

object AudioFormats {
  fun isEncoderSupported(mimeType: String?): Boolean {
    if (mimeType == null) {
      return false
    }
    if (mimeType == MediaFormat.MIMETYPE_AUDIO_RAW) {
      return true
    }

    val mcl = MediaCodecList(MediaCodecList.REGULAR_CODECS)
    for (info in mcl.codecInfos) {
      if (info.isEncoder) {
        for (supportedType in info.supportedTypes) {
          if (supportedType.equals(mimeType, ignoreCase = true)) {
            return true
          }
        }
      }
    }
    return false
  }

  fun getMimeType(encoder: String?): String? {
    return when (encoder) {
      AudioEncoder.AacLc.value, AudioEncoder.AacEld.value, AudioEncoder.AacHe.value -> MediaFormat.MIMETYPE_AUDIO_AAC
      AudioEncoder.AmrNb.value -> MediaFormat.MIMETYPE_AUDIO_AMR_NB
      AudioEncoder.AmrWb.value -> MediaFormat.MIMETYPE_AUDIO_AMR_WB
      AudioEncoder.Wav.value, AudioEncoder.Pcm16bits.value -> MediaFormat.MIMETYPE_AUDIO_RAW
      AudioEncoder.Opus.value -> MediaFormat.MIMETYPE_AUDIO_OPUS
      AudioEncoder.Flac.value -> MediaFormat.MIMETYPE_AUDIO_FLAC
      else -> null
    }
  }
}